clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -1350,21 +1321,33 @@ public Map<Uuid, SortedSet<Integer>> currentAssignment() {
         return this.currentAssignment;
     }
 
-
     /**
      * @return Set of topic IDs received in a target assignment that have not been reconciled yet
-     * because topic names are not in metadata. Visible for testing.
+     * because topic names are not in metadata or reconciliation hasn't finished. Reconciliation
+     * hasn't finished for a topic if the currently active assignment has a different set of partitions
+     * for the topic than the target assignment.
+     *
+     * Visible for testing.
      */
-    Set<Uuid> topicsWaitingForMetadata() {
-        return Collections.unmodifiableSet(assignmentUnresolved.keySet());
+    Set<Uuid> topicsAwaitingReconciliation() {
+        return topicPartitionsAwaitingReconciliation().keySet();
     }
 
     /**
-     * @return Topic partitions received in a target assignment that have been resolved in
-     * metadata and are ready to be reconciled. Visible for testing.
+     * @return Map of topics partitions received in a target assignment that have not been
+     * reconciled yet because topic names are not in metadata or reconciliation hasn't finished.
+     * The value will always contain all partitions in the target assignment.
+     *
+     * Visible for testing.
      */
-    Set<TopicIdPartition> assignmentReadyToReconcile() {
-        return Collections.unmodifiableSet(assignmentReadyToReconcile);
+    Map<Uuid, SortedSet<Integer>> topicPartitionsAwaitingReconciliation() {
+        final Map<Uuid, SortedSet<Integer>> topicPartitionMap = new HashMap<>();
+        currentTargetAssignment.forEach((x, y) -> {
+            if (!currentAssignment.containsKey(x) || !currentAssignment.get(x).equals(y)) {
```

Comment:
The semantic is a bit weird here. The method says that it returns the partitions awaiting reconciliation but the method could also return partitions reconciled if at least one partition of the same topic is not reconciled yet. I suppose that this is what you meant by `The value will always contain all partitions in the target assignment.`.

By the way, it seems that this method is not used in the class. Do we only need it for testing?

Reply:
Yes, it's only used for testing. This method is a replacement for `assignmentReadyToReconcile` which was also used only in testing. You are right that semantics were a bit weird, I implemented so that it replicates the behavior of the (old) internal state. I changed it to have more reasonable semantics (the value contains only the partitions that are missing from the currently reconciled assignment). For the existing tests, there is no difference between the two semantics.